### DISCUSSION TITLE
Add barcode to file name (?)

### DISCUSSION THREAD

**Discussion by frank-from-cologne** [jonaswinkler/paperless-ng#1320] [Support]
Hey,
first of all: I am a paperless-ng user and not a developer.

My wish is to get p-ng to rename an incoming PDF - and to use part of the read content, namely the ASN, for the filename. (*)

The long version:
I put a sticker on the paper documents that has to be scanned, instead of writing down an ASN by hand.
The sticker shows a consecutive which is intended to serve as the document ID, and it shows it twice: as a numerical sequence of digits + as a barcode.
The document is scanned and comes as a PDF in p-ng.
My wish is that this consecutive number (in its function as a document ID) becomes part of the file name, preferably as a kind of suffix, for example: "document.pdf" should become "document-12345678.pdf".
I am aware that - in order for this to work "correctly" in my sense - p-ng must be able to recognise which text string is a document ID and which is just a normal sequence of digits. p-ng could recognise this from the barcode (maybe it can even read it generically)?

Does this option already exist and can I (could a developer) configure p-ng to do this with the existing options? 

Or does anyone see the possibility of developing p-ng in this way? 

I am grateful for every hint and every offer!

Thanks in advance + only the best

Frank

(*) The purpose of this action is to be able to search and find documents in environments that cannot read PDF content, i.e. where I have to rely on the file name.

**Comment by gck303** [comment]
I think that this would be very beneficial. I think that there is a huge gap that a unique machine readable reference on the document (eg a barcode) cannot be automatically got into the p-ng system. 

What you are proposing would be very helpful. Even if teh file was not renamed, it would allow you to find the document!

**Reply by frank-from-cologne** [reply]
What I wonder is – whether and how I can participate in a funding for this function?

**Comment by frank-from-cologne** [comment]
Is it foreseen to "fund" extensions ?
Or to award a sum of money?
Or what do I have to do, how do I have to place my wish so that an interested developer makes me an offer?

**Comment by f-ohler** [comment]
Hi there,

I think we can split this into two separate aspects:

1. parse the ASN from the scanned document for use in paperless-ng
2. adjust the naming schema of your files


Regarding 1: There already was a discussion about this in #460. I adjusted the script to actually parse barcodes instead of relying on the (usually also present) human-readable version of the information, which you may find helpful here. See [this repo](https://github.com/muued/asn-from-barcode)
Regarding 2: [There is a configuration option for this](https://paperless-ng.readthedocs.io/en/latest/advanced_usage.html#file-name-handling). However, the document processing of paperless determines the file names BEFORE calling the post-consume script (at which point the ASN will be empty), thus simply putting both things together will not work. Currently, I don't see a stable way to fix this in a non-brute force manner. The brute-force way would be calling [the document_renamer](https://paperless-ng.readthedocs.io/en/latest/administration.html#managing-filenames) at the end of the post-consume script or in regular intervals. The fragile way would be to rename the files in the post-consume script and adjust the information in the database so that paperless-ng will still find the files. Maybe @jonaswinkler can give us a hint about a better way to do this.

HTH

**Reply by frank-from-cologne** [reply]
Thank you!
So far I can follow.
Asked stupidly (as I said, I'm not a developer) and to understand it: What's wrong with the brute-force way? 

Maybe @jonaswinkler  can actually help us ;-)

**Reply by f-ohler** [reply]
The brute force way triggers the document_renamer which in turn determines the correct name for **every** document in your paperless-ng instance and moves/renames it if needed. As the number of documents in your instance grows, the amount of unnecessary work grows, since most files will not need to be renamed. For instance, if you add that step to the post_consume script, only the freshly added file will need to be renamed, but all other documents are also considered by the document_renamer.

**Comment by sbrunner** [comment]
I'm not certain that, a good idea to use the bar-code/qr-code because more and more documents use qr-code, but it should be relay nice to scan them and add the result in the PDF or in a field of the document :-)

**Reply by frank-from-cologne** [reply]
I see the point. And yes: there could be collisions.
But there are workarounds where it would (still) be absolutely helpful.

**Reply by f-ohler** [reply]
Valid point, absolutely. That's why I added lots of heuristics trying to identify the correct information (filter by barcode type, use closest match to maximum ASN currently stored in DB).
Furthermore, you can add constant prefix/postfix information to your barcodes which can be used by the script to further reduce the chance of situations where multiple barcodes might look relevant and the script would run into troubles identifying the correct ASN.

**Reply by pkoerner81929** [reply]
Hi Frank et all,

before I just started to use Paperless-ng, I had solved that problem through combining/enhancing some existing tools. We may be able to simply adopt that to the consume process of Paperless-ng. 

My solution is as follows:

**1. Barcode sticker**
I ordered similar barcode/number stickers as you, Frank. They use Code128 2D barcodes representing my number scheme, e.g. K120000, K120001, K120002 etc. I am using different sets to separate e.g. private, business etc. I also put a sticker on the 1st page of the document. It could be multiple pages, the sticker goes on page 1.

**2. ADF scanner setup**
I am using a Fujitsu ix500 scanner. I collect multiple documents, e.g. 15-20 pages as a stack to be scanned. 300 DPI, color, front/back pages. The scanner is connected to a Raspberry pi through USB.
 
**3. Scanbutton, scanning, separating, barcode recognition, renaming**
Pressing the scanbutton starts scanning of the whohle stack, resulting in one PDF file (usinig scanbd, sane, scan.sh, compressed etc.,  scripts available). As a next step, I am using a modified pdf-splitter script, zbarimg tools etc. to use a recognized CODE128 barcode with the length of 7 characters as a page split. As a result, I get multiple single files, starting with the recognized Barcode.
E.g. "K120243_Scan_20220119_p1-p3" and "K120244_Scan_20220119_p4-p7" etc.

**4. Sharing documents** 
The documents are shared on a Samba Share to allow access from other devices.

Currently, I manually move the documents to the "consume" folder of paperless-ng.

All of the above is ready to use, works quite stable and may easily be adopted to e.g. different barcode schemes.


**Todo, next steps, idea:**
Let Paperless-ng either use the filename ASN information and populate the ASN field. Or - which I prefer - use e.g. the pre- or post-consume scripts to start the barcode recognintion from above from within paperless-ng to fill the ASN variable. 
I have not tested that yet, so still open to suggestions etc.

Let me know your thoughts,

Peter

**Reply by nm11** [reply]
Hello Peter and all,

Thank you for your detailed answer. It is always good to hear that my own approach is apparently not completely absurd ;-) 

To be quite honest: The "idea" of automatically using a code in the file name, which the software that is doing the OCR pulls from the document, does not come from me, but I have seen this in a software: EXACT SCAN http://exactscan.com, which I used together with a Fuji scanner on my Mac.
But unfortunately, in my workaround it is not possible to capture everything via desktop scanners, for several reasons, e.g. not every document is (well) suited for this, and/or we work with a widely scattered team - most of us are somewhere and use their smartphone and a scan app for capture and submission.

Therefore the wish of having this function in paperless-ng itself.

I would like to remind my purpose behind this request:
As long as I am moving within systems like p-ng, I don't need an ID in the file name, I can search for the file content.
But this ID in the file name would help me if:
- I have the original analogue document and I'm looking for the PDF in an environment that doesn't support a full index (and there are more than I believed, f.e. in the most backups).
- I have the digital document copy and am looking for the original analog doc, e.g. in a decentralised storage structure (It is much less time-consuming for each participant to simply file their original receipts in their own folder or collector in the order of the sticker numbers. I know who has the original by the number circle, and I can find it quickly by the number).

These are my thougts, I am curious of yours :-)

Frank

**Reply by pkoerner81929** [reply]
Actually, I am also a user of Exact Scan (Pro) for years. So, not absurd I guess.

My motivation is to get rid of the need to operate a dedicated mac for just running exact scan. I therefore use the scanbutton to allow others to simple contribute to scan/upload documents and "try" to get the same resulting barcode detection and filename creation. 

Obviously, in PL-ng you would not need to extract the ASN as long it is visible in previews to allow cold storage access. So, I will test the options to leverage pre- and post-consume scripts. Depending on the effort to get the barcode detection in, I may then contribute/share. 

I will keep you posted. 

Peter

Feel free to DM for offline discussion.

**Reply by PhillipRt** [reply]
To make this solution more stable it would be great if you could reserve ASNs in PL-ng. These are then "locked" to the barcode and cannot be handed out to documents manually. 
In the best case this is a button called "Print Barcodes" where a PDF is generated containing the set number of barcodes and these ASNs are locked in the background automatically. 
This would also make it possible to print barcodes with a predefined tags, document types or file names as well as making it easier to check whether a barcode belongs to PL-ng or was printed on by the creator.
